util/module_logger.go

```diff
@@ -0,0 +1,260 @@
+		logs:   make(map[string]log.Logger),
+		l: func(rl *RestrictedLogger) *RestrictedLogger {
+			return rl
+		}}
+	ml.SetLogger(RootModule, logger)
+	return ml.Logger(RootModule).(*ModuleLogger)
+}
+
+type RestrictedLogger struct {
+	logger log.Logger
+	min    Level
+}
+
+func (rl *RestrictedLogger) shouldOutput(level Level) bool { return level >= rl.min }
+func (rl *RestrictedLogger) Debug(args ...interface{}) {
+	if rl.shouldOutput(DebugLevel) {
+		rl.logger.Debug(args...)
+	}
+}
+func (rl *RestrictedLogger) Info(args ...interface{}) {
+	if rl.shouldOutput(InfoLevel) {
+		rl.logger.Info(args...)
+	}
+}
+func (rl *RestrictedLogger) Warn(args ...interface{}) {
+	if rl.shouldOutput(WarnLevel) {
+		rl.logger.Warn(args...)
+	}
+}
+func (rl *RestrictedLogger) Error(args ...interface{}) {
+	if rl.shouldOutput(ErrorLevel) {
+		rl.logger.Error(args...)
+	}
+}
+func (rl *RestrictedLogger) Fatal(args ...interface{}) {
+	if rl.shouldOutput(FatalLevel) {
+		rl.logger.Fatal(args...)
+	}
+}
+func (rl *RestrictedLogger) Panic(args ...interface{}) {
+	if rl.shouldOutput(PanicLevel) {
+		rl.logger.Panic(args...)
+	}
+}
+func (rl *RestrictedLogger) Debugf(format string, args ...interface{}) {
+	if rl.shouldOutput(DebugLevel) {
+		rl.logger.Debugf(format, args...)
+	}
+}
+func (rl *RestrictedLogger) Infof(format string, args ...interface{}) {
+	if rl.shouldOutput(InfoLevel) {
+		rl.logger.Infof(format, args...)
+	}
+}
+func (rl *RestrictedLogger) Warnf(format string, args ...interface{}) {
+	if rl.shouldOutput(WarnLevel) {
+		rl.logger.Warnf(format, args...)
+	}
+}
+func (rl *RestrictedLogger) Errorf(format string, args ...interface{}) {
+	if rl.shouldOutput(ErrorLevel) {
+		rl.logger.Errorf(format, args...)
+	}
+}
+func (rl *RestrictedLogger) Fatalf(format string, args ...interface{}) {
+	if rl.shouldOutput(FatalLevel) {
+		rl.logger.Fatalf(format, args...)
+	}
+}
+func (rl *RestrictedLogger) Panicf(format string, args ...interface{}) {
+	if rl.shouldOutput(PanicLevel) {
+		rl.logger.Panicf(format, args...)
+	}
+}
+func (rl *RestrictedLogger) WithField(key string, value interface{}) *RestrictedLogger {
+	return &RestrictedLogger{
+		logger: rl.logger.WithField(key, value),
+		min:    rl.min}
+}
+func (rl *RestrictedLogger) WithFields(fields log.LogFields) *RestrictedLogger {
+	return &RestrictedLogger{
+		logger: rl.logger.WithFields(fields),
+		min:    rl.min}
+}
+func (rl *RestrictedLogger) Fields() log.Fields {
+	return rl.logger.Fields()
+}
+
+// ModuleLogger is a flat sub-logger configuration and factory. Each sub-logger
+// can be restricted to emit messages above a minimum level or use a different
+// underlying logger. The configuration can be changed on the fly even for
+// existing sub-loggers.
+// Method calls that return other loggers return instead a ModuleLogger
+// instance. This simplifies porting existing code unaware of module logging.
+type ModuleLogger struct {
+	l      func(*RestrictedLogger) *RestrictedLogger
```

Comment:
What is `l`? Can we use a more descriptive name? And/or comment on its purpose might be a good idea.